### PR TITLE
Refactor for better use as external library

### PR DESCRIPTION
**Pull request by ryantking**
Signed-off-by: Ryan King <ryking@redhat.com>

Goal of this PR is to make this repo more usable as an external library.

- [x] Move pinning logic to public package from CLI
- [x] Replace skopeo binary calls with library calls

### CODE REVIEW

File: cmd/pinning/extract.go
Comment (jmrodri) [inline cmd/pinning/extract.go:55]: What about the err from the `pullspec.FromDirectory`. I think we're eating an error somewhere.

File: cmd/pinning/extract.go
Comment (joelanford) [inline cmd/pinning/extract.go:54]: Check this error?

**Review by jmrodri — CHANGES_REQUESTED** [review]
At first glance looks good. There was small place where we seem to be eating an err.

**Comment by zach-source** [comment]
The code looks fine from my perspective, only concern I have is that the integration test failed. It looks like the behavior changed from the previous, there used to be files generated for the pinning command that aren't found any longer. This may not be that big of a deal but it was the behavior of the library before it.

**Review by zach-source — COMMENTED** [review]
Will approve once we can decide on the integration functionality, either fix it or get consensus for removing the file checks there; then it's a merge from me. If no one cares then I'm fine with the files not being generated.

**Comment by ryantking** [comment]
Hi @zach-source that is what we are trying to determine: are the intermediary files part of a compatibility contract that we would be breaking? If so I can add them back in.

**Comment by tonytcampbell** [comment]
Hey folks, while we are not using this yet our plan is to migrate from the python version to this version.  With that said we do rely on those intermediary files at the moment.  I think we could eliminate the files in a future version after a bit of refactoring but would love to see them present for now.

**Comment by zach-source** [comment]
Apologies, I've been MIA due to paternity leave. I'll take a look at making the necessary changes to support the integration features and get this merged. Thanks for the contribution @ryantking

**Comment by ryantking** [comment]
Not a problem @zach-source, and congratulations! I added back the intermediary files to the CLI `pin` command in order to get this PR moved along. Now, this is just a refactoring PR that adds the nuts and bolts to library functions and out of the CLI commands on top of the `crane` resolver, which is opt-in.

**Comment by zach-source** [comment]
@ryantking Thanks Ryan, all test are passing now. I'll just wait for @jmrodri to re-review since they had comments you resolved. I'll merge it as soon as we get a thumbs up.

**Comment by ryantking** [comment]
@zach-source Since this PR really has two components: I rebased the PR into two commits with detailed messages and descriptions. When you decide to merge this PR, I recommend you merge with a commit merge so the individual commits get added.

**Review by jmrodri — APPROVED** [review]

**Comment by ryantking** [comment]
@zach-source After you merge this, can you also cut a new release?

**Comment by zach-source** [comment]
@ryantking Does it automatically.

**Comment by zach-source** [comment]
Well, semiautomaticly. Had to push a tag. Thanks @ryantking

**Pull request merged**
